gk/bpf.c

```diff
@@ -16,13 +16,157 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#define GK_BPF_INTERNAL static
+#include "gatekeeper_flow_bpf.h"
+
 #include "gatekeeper_gk.h"
+#include "gatekeeper_main.h"
+
 #include "bpf.h"
 
 static const struct rte_bpf_xsym flow_handler_init_xsym[] = {
 };
 
+struct gk_bpf_pkt_frame {
+	uint64_t		password;
+	struct flow_entry	*fe;
+	struct rte_mbuf		*pkt;
+	struct gk_config	*gk_conf;
+	struct gk_bpf_pkt_ctx	ctx;
+};
+
+static const uint64_t pkt_password = 0xa2e329ba8b15af05;
+
+static struct gk_bpf_pkt_frame *
+pkt_ctx_to_frame(struct gk_bpf_pkt_ctx *ctx)
+{
+	struct gk_bpf_pkt_frame *frame;
+
+	if (unlikely(ctx == NULL))
+		return NULL;
+
+	frame = container_of(ctx, struct gk_bpf_pkt_frame, ctx);
+	if (unlikely(frame->password != pkt_password)) {
+		GK_LOG(WARNING, "%s(): password violation\n", __func__);
+		return NULL;
+	}
+
+	return frame;
+}
+
+static struct gk_bpf_cookie *
+pkt_ctx_to_cookie(struct gk_bpf_pkt_ctx *ctx)
+{
+	struct gk_bpf_pkt_frame *frame = pkt_ctx_to_frame(ctx);
+	if (unlikely(frame == NULL))
+		return NULL;
+	return &frame->fe->u.bpf.cookie;
+}
+
+static struct rte_mbuf *
+pkt_ctx_to_pkt(struct gk_bpf_pkt_ctx *ctx)
+{
+	struct gk_bpf_pkt_frame *frame = pkt_ctx_to_frame(ctx);
+	if (unlikely(frame == NULL))
+		return NULL;
+	return frame->pkt;
+}
+
+static int
+gk_bpf_encapsulate(struct gk_bpf_pkt_ctx *ctx, int priority)
```

Comment:
It isn't about being enough since both parameters of `gk_bpf_encapsulate()` are passed via registers. `uint8_t` doesn't make it type safe either since priority is at most `PRIORITY_MAX == 63 < 255`. So I prefer to leave it as it is.